file/filesystem_test.go

```diff
@@ -0,0 +1,503 @@
+				TakenTime: &testTime,
+				Extension: "jpg",
+			},
+			expected: filepath.Join("/organized", testTime.Local().Format("2006-01-02"), "file.jpg"),
+			wantErr:  false,
+		},
+		{
+			name: "datetime template",
+			tmpl: "/organized/{meta.taken.datetime}.jpg",
+			meta: &FileMetadata{
+				TakenTime: &testTime,
+				Extension: "jpg",
+			},
+			expected: filepath.Join("/organized", testTime.Local().Format("2006-01-02-15-04-05")+".jpg"),
+			wantErr:  false,
+		},
+		{
+			name: "extension template",
+			tmpl: "/organized/file.{file.extension}",
+			meta: &FileMetadata{
+				Extension: "mp4",
+			},
+			expected: filepath.Join("/organized", "file.mp4"),
+			wantErr:  false,
+		},
+		{
+			name: "camera maker template",
+			tmpl: "/organized/{meta.camera.maker}/file.jpg",
+			meta: &FileMetadata{
+				CameraMaker: "Canon",
+				Extension:   "jpg",
+			},
+			expected: filepath.Join("/organized", "Canon", "file.jpg"),
+			wantErr:  false,
+		},
+		{
+			name: "camera model template",
+			tmpl: "/organized/{meta.camera.model}/file.jpg",
+			meta: &FileMetadata{
+				CameraModel: "EOS 5D",
+				Extension:   "jpg",
+			},
+			expected: filepath.Join("/organized", "EOS 5D", "file.jpg"),
+			wantErr:  false,
+		},
+		{
+			name: "combined template",
+			tmpl: "/organized/{meta.taken.year}/{meta.camera.maker}/{meta.taken.date}.{file.extension}",
+			meta: &FileMetadata{
+				TakenTime:   &testTime,
+				Extension:   "jpg",
+				CameraMaker: "Sony",
+			},
+			expected: filepath.Join("/organized", testTime.Local().Format("2006"), "Sony", testTime.Local().Format("2006-01-02")+".jpg"),
+			wantErr:  false,
+		},
+		{
+			name: "no time data in metadata",
+			tmpl: "/organized/{meta.taken.year}/file.jpg",
+			meta: &FileMetadata{
+				Extension: "jpg",
+			},
+			expected: filepath.Join("/organized", "{meta.taken.year}", "file.jpg"),
+			wantErr:  false,
+		},
+		{
+			name: "normalization applies",
+			tmpl: "/organized/test--file__{file.extension}",
+			meta: &FileMetadata{
+				Extension: "jpg",
+			},
+			expected: filepath.Join("/organized", "test-file_jpg"),
+			wantErr:  false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			result, err := resolveTargetPath(tt.tmpl, tt.meta)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("resolveTargetPath() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if !tt.wantErr && result != tt.expected {
+				t.Errorf("resolveTargetPath() = %q; want %q", result, tt.expected)
+			}
+		})
+	}
+}
+
+func TestScanFiles(t *testing.T) {
+	// Create a temporary directory structure for testing
+	tmpDir := t.TempDir()
+
+	// Create test files
+	testFiles := []string{
+		"image1.jpg",
+		"image2.png",
+		"video1.mp4",
+		"document.txt",
+		"subdir/image3.jpg",
+		"subdir/video2.mkv",
+		"subdir/nested/image4.png",
+	}
+
+	for _, file := range testFiles {
+		fullPath := filepath.Join(tmpDir, file)
+		dir := filepath.Dir(fullPath)
+		if err := os.MkdirAll(dir, 0755); err != nil {
```

Comment:
Use a virtual filesystem instead of creating files on an actual filesystem, it creates a mess.